diff -urN a/include/iwinfo.h b/include/iwinfo.h
--- a/include/iwinfo.h	2018-05-21 20:25:33.000000000 +0530
+++ b/include/iwinfo.h	2018-05-30 10:06:23.272924192 +0530
@@ -98,6 +98,10 @@
 	uint32_t tx_packets;
 	struct iwinfo_rate_entry rx_rate;
 	struct iwinfo_rate_entry tx_rate;
+    char local_lat[32];
+    char local_long[32];
+    char remote_lat[32];
+    char remote_long[32];
 };
 
 struct iwinfo_txpwrlist_entry {
diff -urN a/iwinfo_lua.c b/iwinfo_lua.c
--- a/iwinfo_lua.c	2018-05-21 20:25:33.000000000 +0530
+++ b/iwinfo_lua.c	2018-05-30 10:05:44.945080262 +0530
@@ -293,6 +293,18 @@
 			lua_pushnumber(L, e->tx_rate.rate);
 			lua_setfield(L, -2, "tx_rate");
 
+			lua_pushstring(L, e->local_lat);
+			lua_setfield(L, -2, "local_lat");
+
+			lua_pushstring(L, e->local_long);
+			lua_setfield(L, -2, "local_long");
+
+			lua_pushstring(L, e->remote_lat);
+			lua_setfield(L, -2, "remote_lat");
+
+			lua_pushstring(L, e->remote_long);
+			lua_setfield(L, -2, "remote_long");
+
 			if (e->rx_rate.mcs >= 0)
 			{
 				lua_pushnumber(L, e->rx_rate.mcs);
diff -urN a/iwinfo_qcawifi.c b/iwinfo_qcawifi.c
--- a/iwinfo_qcawifi.c	2018-05-29 16:59:11.000000000 +0530
+++ b/iwinfo_qcawifi.c	2018-05-30 14:22:07.086149686 +0530
@@ -751,6 +751,7 @@
 	uint8_t tmp[24*1024];
 	struct ieee80211req_sta_info *si;
 	struct iwinfo_assoclist_entry entry;
+	char cmd[512];
 
 	if( qcawifi_iswifi(ifname) )
 		return -1;
@@ -789,7 +790,15 @@
 
 			entry.rx_rate.mcs = -1;
 			entry.tx_rate.mcs = -1;
-
+#if 0			
+			snprintf(cmd,512,"echo si->isi_len :%d,tl: %d, l_lat: %s, l_lon :%s,r_lat: %s,r_lon: %s >> /etc/temp",si->isi_len,tl,
+				si->isi_l_latitude, si->isi_l_longitude,si->isi_r_latitude,si->isi_r_longitude);		
+			system(cmd);
+#endif
+			memcpy(entry.local_lat, si->isi_l_latitude+5, sizeof(entry.local_lat));
+			memcpy(entry.local_long, si->isi_l_longitude+5, sizeof(entry.local_long));
+			memcpy(entry.remote_lat, si->isi_r_latitude+5, sizeof(entry.local_long));
+			memcpy(entry.remote_long, si->isi_r_longitude+5, sizeof(entry.remote_long));
 			memcpy(&buf[bl], &entry, sizeof(struct iwinfo_assoclist_entry));
 
 			bl += sizeof(struct iwinfo_assoclist_entry);
