diff -urN libiwinfo-2015-06-01_org/include/iwinfo.h libiwinfo-2015-06-01_mod/include/iwinfo.h
--- libiwinfo-2015-06-01_org/include/iwinfo.h	2018-07-09 13:57:34.289239779 +0530
+++ libiwinfo-2015-06-01_mod/include/iwinfo.h	2019-01-03 12:55:17.562731701 +0530
@@ -102,6 +102,34 @@
     char local_long[32];
     char remote_lat[32];
     char remote_long[32];
+	uint64_t tx_tput;
+	uint64_t rx_tput;
+	uint8_t local_snr_a1;
+	uint8_t local_snr_a2;
+	uint8_t remote_snr_a1;
+	uint8_t remote_snr_a2;
+	uint8_t ip_addr[20];
+	uint32_t local_phy_err;
+	uint32_t remote_phy_err;
+	uint32_t local_mpdu_err;
+	uint32_t remote_mpdu_err;
+	uint32_t local_retries;
+	uint32_t remote_retries;
+    uint8_t l_customer_name[32];
+    uint8_t r_customer_name[32];
+    uint8_t l_link_id;
+    uint8_t r_link_id;
+    int16_t l_noise_floor;
+    int16_t r_noise_floor;
+    uint8_t tx_rate_mcs;
+    uint8_t l_ant_gain;
+    uint8_t r_ant_gain;
+    uint32_t l_bw_limit;
+    uint32_t r_bw_limit;
+    uint8_t l_rtx;
+    uint8_t r_rtx;
+    uint8_t l_tx_power;
+    uint8_t r_tx_power;
 };
 
 struct iwinfo_txpwrlist_entry {
diff -urN libiwinfo-2015-06-01_org/iwinfo_lua.c libiwinfo-2015-06-01_mod/iwinfo_lua.c
--- libiwinfo-2015-06-01_org/iwinfo_lua.c	2018-07-09 13:58:03.333239283 +0530
+++ libiwinfo-2015-06-01_mod/iwinfo_lua.c	2019-01-03 12:55:27.646771234 +0530
@@ -272,6 +272,9 @@
 
 			lua_newtable(L);
 
+			lua_pushstring(L, e->ip_addr);
+			lua_setfield(L, -2, "ip_addr");
+
 			lua_pushnumber(L, e->signal);
 			lua_setfield(L, -2, "signal");
 
@@ -305,6 +308,87 @@
 			lua_pushstring(L, e->remote_long);
 			lua_setfield(L, -2, "remote_long");
 
+			lua_pushnumber(L, e->tx_tput);
+			lua_setfield(L, -2, "tx_tput");
+
+			lua_pushnumber(L, e->rx_tput);
+			lua_setfield(L, -2, "rx_tput");
+
+			lua_pushnumber(L, e->local_snr_a1);
+			lua_setfield(L, -2, "local_snr_a1");
+
+			lua_pushnumber(L, e->local_snr_a2);
+			lua_setfield(L, -2, "local_snr_a2");
+
+			lua_pushnumber(L, e->remote_snr_a1);
+			lua_setfield(L, -2, "remote_snr_a1");
+
+			lua_pushnumber(L, e->remote_snr_a2);
+			lua_setfield(L, -2, "remote_snr_a2");
+
+			lua_pushnumber(L, e->local_phy_err);
+			lua_setfield(L, -2, "local_phy_err");
+
+			lua_pushnumber(L, e->remote_phy_err);
+			lua_setfield(L, -2, "remote_phy_err");
+
+			lua_pushnumber(L, e->local_mpdu_err);
+			lua_setfield(L, -2, "local_mpdu_err");
+
+			lua_pushnumber(L, e->remote_mpdu_err);
+			lua_setfield(L, -2, "remote_mpdu_err");
+
+			lua_pushnumber(L, e->local_retries);
+			lua_setfield(L, -2, "local_retries");
+
+			lua_pushnumber(L, e->remote_retries);
+			lua_setfield(L, -2, "remote_retries");
+
+			lua_pushstring(L, e->l_customer_name);
+			lua_setfield(L, -2, "l_customer_name");
+
+			lua_pushstring(L, e->r_customer_name);
+			lua_setfield(L, -2, "r_customer_name");
+
+			lua_pushnumber(L, e->l_link_id);
+			lua_setfield(L, -2, "l_link_id");
+
+			lua_pushnumber(L, e->r_link_id);
+			lua_setfield(L, -2, "r_link_id");
+
+			lua_pushnumber(L, e->l_noise_floor);
+			lua_setfield(L, -2, "l_noise_floor");
+
+			lua_pushnumber(L, e->r_noise_floor);
+			lua_setfield(L, -2, "r_noise_floor");
+
+			lua_pushnumber(L, e->tx_rate_mcs);
+			lua_setfield(L, -2, "tx_rate_mcs");
+
+			lua_pushnumber(L, e->l_ant_gain);
+			lua_setfield(L, -2, "l_ant_gain");
+
+			lua_pushnumber(L, e->r_ant_gain);
+			lua_setfield(L, -2, "r_ant_gain");
+
+			lua_pushnumber(L, e->l_bw_limit);
+			lua_setfield(L, -2, "l_bw_limit");
+
+			lua_pushnumber(L, e->r_bw_limit);
+			lua_setfield(L, -2, "r_bw_limit");
+
+			lua_pushnumber(L, e->l_rtx);
+			lua_setfield(L, -2, "l_rtx");
+
+			lua_pushnumber(L, e->r_rtx);
+			lua_setfield(L, -2, "r_rtx");
+
+			lua_pushnumber(L, e->l_tx_power);
+			lua_setfield(L, -2, "l_tx_power");
+
+			lua_pushnumber(L, e->r_tx_power);
+			lua_setfield(L, -2, "r_tx_power");
+
 			if (e->rx_rate.mcs >= 0)
 			{
 				lua_pushnumber(L, e->rx_rate.mcs);
diff -urN libiwinfo-2015-06-01_org/iwinfo_qcawifi.c libiwinfo-2015-06-01_mod/iwinfo_qcawifi.c
--- libiwinfo-2015-06-01_org/iwinfo_qcawifi.c	2018-07-09 13:58:28.057238861 +0530
+++ libiwinfo-2015-06-01_mod/iwinfo_qcawifi.c	2019-01-03 12:55:34.946805287 +0530
@@ -799,6 +799,35 @@
 			memcpy(entry.local_long, si->isi_l_longitude+5, sizeof(entry.local_long));
 			memcpy(entry.remote_lat, si->isi_r_latitude+5, sizeof(entry.local_long));
 			memcpy(entry.remote_long, si->isi_r_longitude+5, sizeof(entry.remote_long));
+			entry.tx_tput = si->isi_tx_tput;
+			entry.rx_tput = si->isi_rx_tput;
+			entry.local_snr_a1 = si->isi_local_snr_a1;
+			entry.remote_snr_a1 = si->isi_remote_snr_a1;
+			entry.local_snr_a2 = si->isi_local_snr_a2;
+			entry.remote_snr_a2 = si->isi_remote_snr_a2;
+			memcpy(entry.ip_addr, si->isi_ip_addr, sizeof(si->isi_ip_addr));
+			entry.local_phy_err = si->isi_local_phy_err;
+			entry.remote_phy_err = si->isi_remote_phy_err;
+			entry.local_mpdu_err = si->isi_local_mpdu_err;
+			entry.remote_mpdu_err = si->isi_remote_mpdu_err;
+			entry.local_retries = si->isi_local_retries;
+			entry.remote_retries = si->isi_remote_retries;
+			memcpy(entry.l_customer_name, si->isi_l_customer_name, sizeof(entry.l_customer_name));
+			memcpy(entry.r_customer_name, si->isi_r_customer_name, sizeof(entry.r_customer_name));
+			entry.l_link_id = si->isi_l_link_id;
+			entry.r_link_id = si->isi_r_link_id;
+			entry.l_noise_floor = si->isi_l_noise_floor;
+			entry.r_noise_floor = si->isi_r_noise_floor;
+			entry.tx_rate_mcs = si->isi_kwn_tx_rate_mcs;
+			entry.l_ant_gain = si->isi_l_ant_gain;
+			entry.r_ant_gain = si->isi_r_ant_gain;
+			entry.l_bw_limit = si->isi_l_bw_limit;
+			entry.r_bw_limit = si->isi_r_bw_limit;
+			entry.l_rtx = si->isi_l_rtx;
+			entry.r_rtx = si->isi_r_rtx;
+			entry.l_tx_power = si->isi_l_tx_power;
+			entry.r_tx_power = si->isi_r_tx_power;
+
 			memcpy(&buf[bl], &entry, sizeof(struct iwinfo_assoclist_entry));
 
 			bl += sizeof(struct iwinfo_assoclist_entry);
